src/main.clj

```diff
@@ -1,2 +1,314 @@
+        closers (set (vals pairs)) ;; And here, so we can use contains? later.
+        ;; Create a set by combining the openers and closers.
+        search-set (into openers closers)
+        ;; We can now simply filter the string against the set,
+        ;; To return just the pair-occurrences that we care about.
+        occurrences (filter #(contains? search-set %) s)]
+    (empty? ;; If the height-stack is empty, stuff should be balanced.
+      (reduce (fn [height-stack occurrence]
+                (cond
+                  ;; (contains?) reads better than get or shorthand
+                  (contains? openers occurrence)
+                  (conj height-stack occurrence)
+                  ;; Checks if the closing occurrence closes the top of the height-collecting stack.
+                  ;; If so, pop the top.
+                  ;; (peek) is more efficient than (last) on a vector.
+                  (and (contains? closers occurrence)
+                       (= (get pairs (peek height-stack)) occurrence))
+                  (pop height-stack)
+                  ;; This is sort of a null-case, for dangling closers.
+                  ;; It's efficient to exit-early here, hence wrap in (reduced)
+                  :else (reduced (conj height-stack occurrence))))
+              []
+              occurrences))))
+; I wanted to solve this by regex pattern matching,
+; Realized that iterating over is much more efficient.
+
+; https://4clojure.oxal.org/#/problem/76
+; Writing trampoline is easy
+; Check if return value is funciton using function?
+; Do recursion accordingly.
+(= [1 3 5 7 9 11]
+   (letfn
+     [(foo [x y] #(bar (conj x y) y))
+      (bar [x y] (if (> (last x) 10)
+                   x
+                   #(foo x (+ 2 y))))]
+     (trampoline foo [] 1)))
+
+; https://4clojure.oxal.org/#/problem/106
+; Double, Halve or Add 2
+; Unsolved
+
+; https://4clojure.oxal.org/#/problem/73
+; This solution isn't extensible :( 😭
+((fn tic-tac-toe
+   ([board] (tic-tac-toe 0 board))
+   ([c board]
+    ; Iterate over all rows to find a winner.
+    (let [winner (reduce (fn [_ r]
+                           (when (and (= 1 (count (distinct r)))
+                                      (not (= :e (first r))))
+                             (reduced (first r))))
+                         []
+                         board)]
+      (if winner
+        winner
+        (case c
+          ; Transpose the matrix over diagonal so columns become rows.
+          0 (tic-tac-toe 1 (apply map vector board))
+          ; Provide 2 diagonals as rows.
+          1 (tic-tac-toe 2 [(map-indexed (fn [i e] (get e i)) board)
+                            (map-indexed (fn [i e] (get e (- 2 i))) board)])
+          2 nil)))))
+ [[:x :e :e]
+  [:o :x :e]
+  [:o :e :x]])
+
+; https://adventofcode.com/2021/day/1
+((defn aoc1-part1
+   [file]
+   (with-open [rdr (clojure.java.io/reader file)]
+     (:count
+       (reduce
```

Comment:
Instead of large reducers, try and break the problem down into smaller units. Think pipelines not procedures. Take some input, process it, pass it over, transform it, repeat.

This problem can be much simply solved by just comparing the last value of the current step against the first element of the previous step since the middle values are the same. Since the rolling window has repeating elements.

```
[1 3 5]
[3 5 7]
[5 7 9]
```

Comparing 1 with 7, 3 with 9 etc.